sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/it/SpannerChangeStreamOrderedByTimestampAndTransactionIdIT.java

```diff
@@ -0,0 +1,659 @@
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.PipelineResult;
+import org.apache.beam.sdk.coders.BooleanCoder;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerConfig;
+import org.apache.beam.sdk.io.gcp.spanner.SpannerIO;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.DataChangeRecord;
+import org.apache.beam.sdk.io.gcp.spanner.changestreams.model.Mod;
+import org.apache.beam.sdk.state.BagState;
+import org.apache.beam.sdk.state.StateSpec;
+import org.apache.beam.sdk.state.StateSpecs;
+import org.apache.beam.sdk.state.TimeDomain;
+import org.apache.beam.sdk.state.Timer;
+import org.apache.beam.sdk.state.TimerSpec;
+import org.apache.beam.sdk.state.TimerSpecs;
+import org.apache.beam.sdk.state.ValueState;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * End-to-end test of Cloud Spanner Change Streams with strict commit timestamp and transaction
+ * ordering.
+ */
+@RunWith(JUnit4.class)
+public class SpannerChangeStreamOrderedByTimestampAndTransactionIdIT {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SpannerChangeStreamOrderedByTimestampAndTransactionIdIT.class);
+
+  @ClassRule public static final IntegrationTestEnv ENV = new IntegrationTestEnv();
+  @Rule public final transient TestPipeline pipeline = TestPipeline.create();
+
+  private static String projectId;
+  private static String instanceId;
+  private static String databaseId;
+  private static String tableName;
+  private static String changeStreamName;
+  private static DatabaseClient databaseClient;
+
+  @BeforeClass
+  public static void setup() throws InterruptedException, ExecutionException, TimeoutException {
+    projectId = ENV.getProjectId();
+    instanceId = ENV.getInstanceId();
+    databaseId = ENV.getDatabaseId();
+    tableName = ENV.createSingersTable();
+    changeStreamName = ENV.createChangeStreamFor(tableName);
+    databaseClient = ENV.getDatabaseClient();
+  }
+
+  @Test
+  public void testTransactionBoundaries() {
+    final SpannerConfig spannerConfig =
+        SpannerConfig.create()
+            .withProjectId(projectId)
+            .withInstanceId(instanceId)
+            .withDatabaseId(databaseId);
+    // Commit a initial transaction to get the timestamp to start reading from.
+    List<Mutation> mutations = new ArrayList<>();
+    mutations.add(insertRecordMutation(0, "FirstName0", "LastName0"));
+    final long timeIncrementInSeconds = 5;
```

Comment:
Could we reduce this time or even better not have it? If not, could we write some code to check for a condition in a loop and sleep on smaller intervals until a timeout is reached?

I ask this, because introducing sleeps adds an artificial lower bound to how fast this test can be, which penalizes the whole integration test suite performance.

Reply:
Sure, reduced the sleep time to 2 seconds.

I think the sleep time is necessary, since i'm testing that the buffer flushes at advancing timestamps.